### PR TITLE
Canwan/lowdb to etcd

### PR DESCRIPTION
**Pull request by wangcan0329**
[User Authentication] remove lowdb dependency change to etcd2

### CODE REVIEW

File: service-deployment/bootstrap/rest-server/rest-server.yaml.template [hunk 1]
```diff
@@ -48,12 +48,12 @@ spec:
           value: {{ clusterinfo['restserverinfo']['webhdfs_uri'] }}
         - name: JWT_SECRET
           value: {{ clusterinfo['restserverinfo']['jwt_secret'] }}
-        - name: LOWDB_FILE
-          value: /var/pai/{{ clusterinfo['restserverinfo']['lowdb_path'] }}
-        - name: LOWDB_ADMIN
-          value: "{{ clusterinfo['restserverinfo']['lowdb_admin'] }}"
-        - name: LOWDB_PASSWD
-          value: "{{ clusterinfo['restserverinfo']['lowdb_passwd'] }}"
+        - name: ETCD_ADMIN
```
Comment (fanyangCS) [inline service-deployment/bootstrap/rest-server/rest-server.yaml.template]: >ETCD_ADMIN [](start = 16, length = 10)

is this the default user/password? maybe to rename it as PAI_ADMIN is better?

File: service-deployment/clusterconfig-example.yaml [hunk 2]
```diff
@@ -80,12 +80,13 @@ clusterinfo:
...
     # secret for signing authentication tokens, e.g., "Hello PAI!"
     jwt_secret: your_jwt_secret
-    # authentication database file path relative to PAI data path
-    lowdb_path: "rest-server/user.db.json"
-    # database admin username
-    lowdb_admin: your_admin_username
-    # database admin password
-    lowdb_passwd: your_admin_password
+    # etcd database admin username
+    etcd_admin: your_admin_username
```
Comment (fanyangCS) [inline service-deployment/clusterconfig-example.yaml]: >etcd_admin [](start = 4, length = 10)

Better to rename it as pai_admin?

File: rest-server/test/setup.js [hunk 3]
```diff
@@ -23,9 +23,9 @@ process.env.HDFS_URI = 'hdfs://hdfs.test.pai:9000';
 process.env.WEBHDFS_URI = 'http://hdfs.test.pai:50070';
 process.env.LAUNCHER_WEBSERVICE_URI = 'http://launcher.test.pai:9086';
 process.env.JWT_SECRET = 'jwt_test_secret';
-process.env.LOWDB_FILE = './user.db.json';
-process.env.LOWDB_ADMIN = 'iamadmin';
-process.env.LOWDB_PASSWD = 'adminisi';
+process.env.ETCD_URI = 'http://etcd.test.ip1.pai:4001,http://etcd.test.ip2.pai:4001';
```
Comment (fanyangCS) [inline rest-server/test/setup.js]: >ETCD_URI [](start = 12, length = 8)

I did not see any UT regarding to add/remove user. can you add the corresponding UTs?

File: rest-server/src/util/etcd2.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+// module dependencies
+const Etcd = require('node-etcd');
+const StorageBase = require('./storageBase');
+
+class Etcd2 extends StorageBase {
```
Comment (fanyangCS) [inline rest-server/src/util/etcd2.js]: >Etcd2 [](start = 6, length = 5)

please add UTs for get/set/has/delete

Comment (wangcan0329) [reply rest-server/src/util/etcd2.js]: I will push UT to this pr later

File: service-deployment/clusterconfig-example.yaml (hunk 2 above)
Comment (hwuu) [reply service-deployment/clusterconfig-example.yaml]: Agree with Fan. 'admin' is PAI's admin, not etcd's. And so is the 'password'. Suggest change to:
- `etcd_admin` --> `default_pai_admin_username`, and
- `etcd_passwd` --> `default_pai_admin_password`.

File: rest-server/src/util/etcd2.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  constructor(options) {
+    super();
+    this.options = options;
+    this.etcdClient = new Etcd(this.options.hosts);
+  }
+
+  get(key, callback, options = null) {
```
Comment (hwuu) [inline rest-server/src/util/etcd2.js]: Call back functions are usually put behind all other parameters in the parameter list. Please take this order:
`key` --> `options` --> `callback`.

File: rest-server/src/util/etcd2.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    try {
+      this.etcdClient.get(key, options, (err, res) => {
+        if (err === null) {
+          let kvMap = this.flattenRes(res.node);
```
Comment (hwuu) [inline rest-server/src/util/etcd2.js]: What does `flattenRes` do?

File: rest-server/src/models/user.js [hunk 7]
```diff
@@ -37,49 +36,126 @@ const encrypt = (username, password, callback) => {
...
-          db.set(`${username}.admin`, admin).write();
-        }
-        callback(null, true);
+        encrypt(username, password, (errMsg, derivedKey) => {
+          if (errMsg) {
+            callback(errMsg, false);
+          } else {
+            if (modify) {
+              db.set(etcdConfig.userPasswdPath(username), derivedKey, (errMsg, res) => {
+                if (errMsg !== null) {
```
Comment (abuccts) [inline rest-server/src/models/user.js]: use `if (errMsg)`? pls refer to [difference between null and undefined](https://stackoverflow.com/a/5101991/7630831)

File: rest-server/src/util/etcd2.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  constructor(options) {
+    super();
+    this.options = options;
+    this.etcdClient = new Etcd(this.options.hosts);
+  }
+
+  get(key, callback, options = null) {
+    try {
+      this.etcdClient.get(key, options, (err, res) => {
+        if (err === null) {
```
Comment (abuccts) [inline rest-server/src/util/etcd2.js]: same, use `if (!err)`?

File: rest-server/src/models/user.js (hunk 7 above)
Comment (wangcan0329) [reply rest-server/src/models/user.js]: Got it! I will change to if (errMsg). Thanks

**Pull request merged**